src/execution.rs

```diff
@@ -114,6 +122,116 @@ pub trait Report {
     fn start(&mut self, task_name: String);
     fn report(&mut self, result: &CompletedTask);
     fn done(&mut self);
+
+    fn stage(&mut self, full_name: &[String], stage_rep: StageReport) {
+        let mut full_name: Vec<String> = Vec::from(full_name);
+        full_name.push(stage_rep.stage_name);
+        let completed_task = CompletedTask {
+            full_name,
+            duration: stage_rep.duration,
+            stdout: Vec::new(),
+            stderr: Vec::new(),
+            status: Status::from(stage_rep.status),
+        };
+        self.report(&completed_task);
+    }
+}
+
+pub struct StageReportSender {
+    sender: pipe::Sender,
+    // Since we define the stages to be linear, we just need to
+    // keep one timestamp to report a stage's duration.
+    started_at: Instant,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub enum StageStatus {
+    Success,
+    Failure(i32),
+    Skipped(String),
+}
+
+impl From<StageStatus> for Status {
+    fn from(s: StageStatus) -> Self {
+        match s {
+            StageStatus::Success => Status::Success,
+            StageStatus::Failure(code) => Status::Failure(code),
+            StageStatus::Skipped(reason) => Status::Skipped(reason),
+        }
+    }
+}
+
+#[derive(PartialEq, Debug, Serialize, Deserialize)]
+pub struct StageReport {
+    stage_name: String,
+    status: StageStatus,
+    duration: Duration,
+}
+
+impl StageReportSender {
+    pub fn report_stage_status<N: ToString>(&mut self, stage_name: N, status: StageStatus) {
+        let stage_name = stage_name.to_string();
+        let end = Instant::now();
+        let start = self.started_at;
+        self.started_at = end;
+
+        let payload = StageReport {
+            stage_name,
+            status,
+            duration: end.duration_since(start),
+        };
+
+        serialize_and_write(&mut self.sender, &payload).expect("Couldn't send");
+    }
+}
+
+fn serialize_and_write<W: Write, A: Serialize>(w: &mut W, payload: &A) -> io::Result<usize> {
+    let payload = bincode::serialize(payload).unwrap();
+    let n = w.write(&(payload.len() as usize).to_be_bytes())?;
+    let m = w.write(&payload)?;
+    Ok(n + m)
+}
+
+struct StreamDecoder {
+    buf: Vec<u8>,
+    offset: usize,
+}
+
+impl StreamDecoder {
+    fn new() -> Self {
+        StreamDecoder {
+            buf: Vec::new(),
+            offset: 0,
+        }
+    }
+
+    fn append(&mut self, data: &[u8]) {
+        self.buf.extend_from_slice(data);
+    }
+
+    // Decode a message if there is enough data in the buffer.
+    fn try_decode(&mut self) -> Option<StageReport> {
+        let avail = self.buf.len() - self.offset;
+
+        if avail <= size_of::<usize>() {
```

Comment:
```suggestion
        if avail < size_of::<usize>() {
```

I'd assume a default value can be deserialized as an empty byte sequence…